tests/test_group_by.cc

```diff
@@ -471,6 +473,40 @@ TEST(test_groupby, test_group_by_first_group) {
   EXPECT_EQ(groups.first_value(), first_group.second);
 }
 
+TEST(test_groupby, test_group_by_first_last_value) {
+
+  const auto fib = fibonacci(20);
+
+  const auto grouped_indexers = group_by(fib, number_of_digits).indexers();
+  const auto smallest_number_of_digits_inds = grouped_indexers.first_value();
+  const auto largest_number_of_digits_inds = grouped_indexers.last_value();
+  auto it_min =
+      std::distance(fib.begin(), std::min_element(fib.begin(), fib.end()));
+  auto it_max =
+      std::distance(fib.begin(), std::max_element(fib.begin(), fib.end()));
+  unsigned long it_min_u_l = (unsigned long)it_min;
+  unsigned long it_max_u_l = (unsigned long)it_max;
```

Comment:
I think that this is checking that `first_value()` returns the value associated with the minimum key ... which works out here because internally we're using `std::map` which sorts by key, but I _think_ I would have expected `first_value()` and `last_value()` to simply return the first and last elements in the container (sorted or not).

Basically, I think if we changed the container to `std::unordered_map<>` instead of `std::map` these tests could fail even if `first_value()` and `last_value()` continued to do the expected thing.
```suggestion
const auto grouped = group_by(fib, number_of_digits);
const unsigned long expected_first = grouped.begin()->second;
const unsigned long expected_last = std::prev(grouped.end())->second;
EXPECT_EQ(expected_first, grouped.first_value());
EXPECT_EQ(expected_last, grouped.last_value());
```

Reply:
first_value and last_value return all values assigned to the first and last key respectively. So in my case, which is based on the use case, the indexer has keys for each time, sorted by time. And the map has all indices for those entries that have those times. I agree we are also testing the indexer here, but that was actually my point based on the use case.

Reply:
The test also tests that all the right indices are chosen from fib. That the first and last value vectors are complete.